llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -0,0 +1,352 @@
+  }
+  MinTripCount = *ParsedMinTripCount;
+  if (MinTripCount < 2) {
+    LLVM_DEBUG(dbgs() << " PostPipeliner: min tripcount < 2\n");
+    return false;
+  }
+
+  return true;
+}
+
+// This assigns Cycle of SU, Earliest of its predecessors and Earliest of
+// the next instance of SU.
+void PostPipeliner::scheduleNode(SUnit &SU, int Cycle) {
+  LLVM_DEBUG(dbgs() << "PostPipeline " << SU.NodeNum << " in cycle " << Cycle
+                    << "\n");
+  Info[SU.NodeNum].Cycle = Cycle;
+  for (auto &Dep : SU.Succs) {
+    int Latency = Dep.getSignedLatency();
+    auto *Succ = Dep.getSUnit();
+    if (Succ->isBoundaryNode()) {
+      continue;
+    }
+    int SNum = Succ->NodeNum;
+    Info[SNum].Earliest = std::max(Info[SNum].Earliest, Cycle + Latency);
+    LLVM_DEBUG(dbgs() << "   Earliest(" << SNum << ") to "
+                      << Info[SNum].Earliest << "\n");
+  }
+
+  int Next = SU.NodeNum + NInstr;
+  if (Next < NTotalInstrs) {
+    Info[Next].Earliest = std::max(Info[Next].Earliest, Cycle + II);
+  }
+}
+
+// Check resources. We only insert at the position modulo II. Since we insert
+// all iterations separately, the resources that wrap around accumulate in the
+// overflow area, causing conflicts when inserting future iterations
+int PostPipeliner::fit(MachineInstr *MI, int Earliest, int NTries, int II) {
+  for (int C = Earliest; C < Earliest + NTries; C++) {
+    int Mod = C % II;
+    LLVM_DEBUG(dbgs() << "   at " << C << " (" << Mod << ")\n");
+    if (!HR.checkConflict(Scoreboard, *MI, -Depth + Mod)) {
+      LLVM_DEBUG(dbgs() << "    Success\n");
+      return C;
+    }
+  }
+  LLVM_DEBUG(dbgs() << "    Fail\n");
+
+  return -1;
+}
+
+void PostPipeliner::computeLoopCarriedEarliest() {
+  // We schedule the first iteration, only using earliest. This updates
+  // earliest of the successors. Any successor in the second iteration
+  // represents a loop carried dependence, and we account for that by
+  // propagating its Earliest back to the first iteration
+  // Note that we don't have to clean the effects of this exploration,
+  // since the real scheduling will overwrite Cycle, and the ultimate Earliest
+  // will never be less than we compute here.
+  for (int K = 0; K < NInstr; K++) {
+    SUnit &SU = DAG->SUnits[K];
+    const int Earliest = Info[K].Earliest;
+    scheduleNode(SU, Earliest);
+  }
+  for (int K = 0; K < NInstr; K++) {
+    int K2 = K + NInstr;
+    const int Earliest = Info[K2].Earliest - II;
+    Info[K].Earliest = std::max(Info[K].Earliest, Earliest);
+  }
+}
+
+bool PostPipeliner::scheduleFirstIteration() {
+  // Set up the basic schedule from the original instructions
+  for (int K = 0; K < NInstr; K++) {
+    const int N = K;
+    LLVM_DEBUG(dbgs() << "  Trying " << N << "\n");
+    SUnit &SU = DAG->SUnits[N];
+    MachineInstr *MI = SU.getInstr();
+    const int Earliest = Info[N].Earliest;
+    // The first iteration should fit entirely within II
+    const int Actual = fit(MI, Earliest, II, II);
```

Comment:
Oh, I took that comment too literally, obviously we do not want the whole thing to fit within II, otherwise that would be a single stage. I don't think I understand the comment then. What are we protecting ourselves against here?

Reply:
Ah. This comment has some history. fit() actually tries to fit in Earliest % II. If that wraps around, it effectively moves one stage further. So NTries = II just says "Try it in every position in the modulo schedule". More than that is useless.